src/Microsoft.Health.Fhir.SqlServer/Features/Watchdogs/InvisibleHistoryCleanupWatchdog.cs

```diff
@@ -42,40 +42,46 @@ internal InvisibleHistoryCleanupWatchdog()
-
-            IReadOnlyList<(long TransactionId, DateTime? VisibleDate, DateTime? InvisibleHistoryRemovedDate)> transToClean =
-                await _store.GetTransactionsAsync(lastTranId, visibility, cancellationToken, DateTime.UtcNow.AddDays(-1 * RetentionPeriodDays));
-
-            _logger.LogInformation($"{Name}: found transactions={transToClean.Count}.");
-
-            if (transToClean.Count == 0)
-            {
-                _logger.LogDebug($"{Name}: completed. transactions=0.");
-                return;
-            }
-
+            _logger.LogInformation($"{Name}: visibility={visibility}.");
             var totalRows = 0;
-            foreach ((long TransactionId, DateTime? VisibleDate, DateTime? InvisibleHistoryRemovedDate) tran in
-                     transToClean.Where(x => !x.InvisibleHistoryRemovedDate.HasValue).OrderBy(x => x.TransactionId))
+            var totalTrans = 0;
+            var transToCleanCount = 0;
+            do
             {
-                var rows = await _store.MergeResourcesDeleteInvisibleHistory(tran.TransactionId, cancellationToken);
-                _logger.LogInformation($"{Name}: transaction={tran.TransactionId} removed rows={rows}.");
-                totalRows += rows;
-
-                await _store.MergeResourcesPutTransactionInvisibleHistoryAsync(tran.TransactionId, cancellationToken);
+                var lastTranId = await GetLastCleanedUpTransactionIdAsync(cancellationToken);
+                _logger.LogInformation($"{Name}: last cleaned up transaction={lastTranId}.");
+                var transToClean = await _store.GetTransactionsAsync(lastTranId, visibility, cancellationToken, DateTime.UtcNow.AddDays(-1 * RetentionPeriodDays));
+                _logger.LogInformation($"{Name}: found transactions={transToClean.Count}.");
+                if (transToClean.Count > 0)
+                {
+                    foreach (var tran in transToClean.Where(x => !x.InvisibleHistoryRemovedDate.HasValue).OrderBy(x => x.TransactionId))
+                    {
+                        var rows = await _store.MergeResourcesDeleteInvisibleHistory(tran.TransactionId, cancellationToken);
+                        _logger.LogInformation($"{Name}: transaction={tran.TransactionId} removed rows={rows}.");
+                        totalRows += rows;
+                        totalTrans++;
+
+                        await _store.MergeResourcesPutTransactionInvisibleHistoryAsync(tran.TransactionId, cancellationToken);
+                    }
+
+                    var cleanedUpToTranId = transToClean.Max(_ => _.TransactionId);
+                    await UpdateLastCleanedUpTransactionId(cleanedUpToTranId);
+                    _logger.LogInformation($"{Name}: cleaned up to {cleanedUpToTranId}. transactions={totalTrans} removed rows={totalRows}");
+                }
+
+                transToCleanCount = transToClean.Count;
             }
+            while (transToCleanCount > 0);
```

Comment:
Do we need any kind of limit on this loop?  We always want it to run through all transactions, no matter the number?

Reply:
Yes we do. I don't see the need to limit.